grammar/macro.g

```diff
@@ -0,0 +1,13 @@
+MacroCall = path:Path "!" ident_input:IDENT? input:MacroInput;
```

Comment:
```rust
macro_rules! name {
    // ....
}
```

It depends on whether `macro_rules` is considered a `MacroCall`, I suppose. As far as I'm aware, only `macro_rules!` has this syntax.

Reply:
Hm, interesting.

What I was thinking of is that foo!bar{} is considered a valid statement in libsyntax (https://github.com/rust-lang/rust/blob/65e485d8f1d28102b426c9d6d82f835cd6470d3e/src/libsyntax/parse/parser.rs#L4711-L4714).  I assumed that was what it was referring to.  `macro_rules` is parsed through a separate mechanism (`eat_macro_def`).  This "special identifier" was added 6 years ago, and I couldn't tell from [the commit](https://github.com/rust-lang/rust/commit/bd92499c5e65248b3dd2078e2f428f33004cba3d) what it means.  All other instances of parsing macro invocations do not check for that identifier.

Reply:
I think I'm understanding it a little better.  However, macro calls are parsed differently in various situations, and some don't allow the extra ident (type, pattern, trait/impl item, foreign item).  Eventually these differences will need to be captured in the grammar, right?

Also, I think the syntax for `macro_rules` should be captured somewhere, since it is useful for certain rust tooling, syntax highlighting, etc.  It is now in the reference, but I'm hoping the wg-grammar and the reference can be closely tied if not identical.  Maybe the reference will just have extra stuff beyond the wg-grammar?

Reply:
> Also, I think the syntax for `macro_rules` should be captured somewhere

We could do this, especially since `macro` definitions are not just macro invocations, and we could even have a converter from `macro_rules` input patterns to CFGs.

Reply:
> Eventually these differences will need to be captured in the grammar, right?

Ideally we would find a way to keep adding rules without making the grammar unreadable *and* without losing the ability to encode into (E)BNF (automatically).